processing/src/main/java/org/apache/druid/segment/join/JoinMatcher.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.segment.join;
+
+import org.apache.druid.segment.ColumnSelectorFactory;
+
+/**
+ * An interface returned by {@link Joinable#makeJoinMatcher} and used by {@link HashJoinEngine} to implement a join.
+ *
+ * A typical usage would go something like:
+ *
+ * <pre>
+ * matcher.matchCondition();
+ * while (matcher.hasMatch()) {
+ *   // Do something with the match
+ *   matcher.nextMatch();
+ * }
+ * </pre>
+ */
+public interface JoinMatcher
+{
+  /**
+   * Returns a factory for reading columns from the {@link Joinable} that correspond to matched rows.
+   */
+  ColumnSelectorFactory getColumnSelectorFactory();
+
+  /**
+   * Matches against the {@link ColumnSelectorFactory} and {@link JoinConditionAnalysis} supplied to
+   * {@link Joinable#makeJoinMatcher}.
+   *
+   * After calling this method, {@link #hasMatch()} will return whether at least one row matched. After reading that
+   * row, {@link #nextMatch()} can be used to move on to the next row.
+   */
+  void matchCondition();
+
+  /**
+   * Matches every row that has not already been matched. Used for right joins.
+   *
+   * After calling this method, {@link #hasMatch()} will return whether at least one row matched. After reading that
+   * row, {@link #nextMatch()} can be used to move on to the next row.
+   *
+   * Will only work correctly if {@link Joinable#makeJoinMatcher} was called with {@code remainderNeeded == true}.
+   */
+  void matchRemainder();
```

Comment:
Suggest adding some more info here about the relationship/ordering between `matchCondition()` and `matchRemainder()` calls, e.g., is it an error if `matchCondition` is called after `matchRemainder`? (looking at how it's used in `HashJoinEngine.matchCurrentPosition`)

Reply:
Hmm, the way they're currently specced and implemented, they will work no matter what order you call them in. I think this is fine. It's more generic than necessary but I don't think it hurts.